pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.document;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.StringReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+
+import junitparams.JUnitParamsRunner;
+import junitparams.Parameters;
+
+@RunWith(JUnitParamsRunner.class)
+public class TextFileContentTest {
+
+    // in real life it's System.lineSeparator()
+    // to make the class more testable (and avoid some test failures being hidden depending on the platform),
+    // we use this dummy value
+    private static final String LINESEP_SENTINEL = ":fallback:";
+
+    @Rule
+    public ExpectedException expect = ExpectedException.none();
+
+    @Test
+    @Parameters(source = TextContentOrigin.class)
+    public void testMixedDelimiters(TextContentOrigin origin) throws IOException {
+        TextFileContent content = origin.normalize("a\r\nb\n\rc");
+        Assert.assertEquals(Chars.wrap("a\nb\n\nc"), content.getNormalizedText());
+        Assert.assertEquals(LINESEP_SENTINEL, content.getLineTerminator());
+    }
+
+    @Test
+    @Parameters(source = TextContentOrigin.class)
+    public void testFormFeedIsNotNewline(TextContentOrigin origin) throws IOException {
+        TextFileContent content = origin.normalize("a\f\nb\nc");
+        Assert.assertEquals(Chars.wrap("a\f\nb\nc"), content.getNormalizedText());
+        Assert.assertEquals("\n", content.getLineTerminator());
+    }
+
+    @Test
+    public void testNormTextPreservation() {
+        Chars input = Chars.wrap("a\nb\nc");
+        TextFileContent content = TextFileContent.fromCharSeq(input);
+        Assert.assertSame(input, content.getNormalizedText());
+        Assert.assertEquals("\n", content.getLineTerminator());
+    }
+
+    @Test
+    @Parameters(source = TextContentOrigin.class)
+    public void testBomElimination(TextContentOrigin origin) throws IOException {
+        TextFileContent content = origin.normalize("\ufeffabc");
+        Chars normalizedText = content.getNormalizedText();
+        Assert.assertEquals(Chars.wrap("abc"), normalizedText);
+        // this means the underlying string does not start with the bom marker
+        // it's useful for performance to have `textDocument.getText().toString()` be O(1).
+        Assert.assertSame(normalizedText.toString(), normalizedText.toString());
```

Comment:
I think both would be identical. The thing this is testing is that the Chars instance `normalizedText` has the same size as the underlying string, therefore, repeated application of toString will return that string instead of calling substring on the underlying string.

Reply:
I've tried to clarify this in comments